1단계 - 질문 삭제하기 기능 리팩토링

kangbada0728 opened the pull request:
제가 느끼기에 이 미션에서 단위 테스트하기 어려운 부분은 Question 삭제 테스트 시 Answer 까지 제대로 삭제되는지 확인하는 부분이었습니다.
Answer 가 JPA(OneToMany)로 Question 과 묶여있다 보니 Question 생성시 Answer 를 함께 넣어줄 수가 없었기에 때문입니다. 그래서 Question 삭제 테스트는 단순히 소유자의 삭제여부만 확인하면서 수행할 수 밖에 없었습니다.

그리고 나머지 부분은 딱히 테스트하기 어려운 부분은 없었다고 생각합니다.

제가 이렇게 생각한 것이 옳은지, 아니면 다른 해석이 있는지 궁금합니다.

감사합니다.

kangbada0728 committed:
refactor: 비즈니스 핵심 로직을 도메인으로 옮김

kangbada0728 committed:
refactor: 공통 상위 엔티티를 이용하여 인스턴스 변수의 개수를 줄임

csh0034 commented on src/main/java/qna/service/QnAService.java:
```diff
@@ -35,24 +35,6 @@ public Question findQuestionById(Long id) {
     @Transactional
     public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteException {
```
비지니스로직을 도메인으로 옮겨 리팩토링 잘해주셨네요 👍

csh0034 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -23,38 +21,17 @@ public class Question extends AbstractEntity {
     @OrderBy("id ASC")
     private List<Answer> answers = new ArrayList<>();
```
이부분에 추가해주신 일급컬렉션을 적용해주시면 좋을것 같습니다.

```java
@Embedded
private Answers answers = new Answers();
```

kangbada0728 replied:
제가 여기에 일급컬렉션 Answers 를 적용하지 않은 이유는 아직 JPA가 익숙치 않아서 여기에 Answers를 적용해도 @OneToMany 가 제대로 작동하는지 모르겠어서 였습니다. 적용해도 문제가 없을까요?

csh0034 replied:
간단한 샘플프로젝트를 만드셔서 테스트 해보시고 적용해보시는것도 좋을것 같습니다 😃

csh0034 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,24 @@
+        this.answers = answers;
+    }
+
+    public List<DeleteHistory> delete(User deleteBy) throws CannotDeleteException {
```
사다리 미션에서 학습하셨던 Stream api 를 적용해보시면 좋을것 같습니다.

요구사항에 있는 내용은 아니지만 현재 CannotDeleteException 의 경우 Checked Exception 이므로
throws 또는 catch 가 반드시 필요한데요 Unchecked Exception 으로 처리해보는것도 좋을것 같습니다.

kangbada0728 replied:
CannotDeleteException을 Unchecked Exception으로 바꾸기 위해 Exception 대신 RuntimeException을 상속하도록 변경하였습니다. 다른 예외들은 RuntimeException을 상속하여 구현되어 있었고 이 클래스만 Exception을 상속하여 구현되어 있었기에 왜 굳이 이것만 이렇게 구현하셨을까 고민해보았지만, 제가 느끼기에 RuntimeException으로 변경하여도 문제가 없다고 느껴서 변경하였습니다.

csh0034 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,75 +1,70 @@
     }
 
+    public DeleteHistory delete(User deleteBy) throws CannotDeleteException {
+        if (!isOwner(deleteBy)) {
```
검증 추가 👍

csh0034 commented on src/main/java/qna/domain/ArticleEntity.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashCode 를 override 해야하는 이유가 있을까요? 🤔

csh0034 commented on src/main/java/qna/domain/ArticleEntity.java:
```diff
@@ -0,0 +1,56 @@
+
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public class ArticleEntity extends UpdatedEntity {
```
중복제거를 위해 상속 용도로 추가해주셨으니 abstract 키워드를 추가해주시는건 어떨까요?

하단부분은 적용은 꼭 안하셔도되고 제 생각만 공유드려요!

저는 Auditing 기능을 제외하곤 중복이 있더라도 따로 상속용 클래스를 만들지 않는데요
혹시 나중에 구현클래스에서 필요한 필드가 제외 될 경우 상속 클래스도 변경을 해야하기
때문입니다.

또한 deleted 필드는 보통 soft delete 용도로 사용하기 때문에 꼭 Article 과 관련이 있는부분은
아니기 때문에 contents 만 남는다면 기존처럼 각 구현 클래스에 두어도 될것 같습니다.

kangbada0728 replied:
말씀 감사합니다. 저도 이렇게까지 하는게 옳은가 의문이 있었지만, 한 클래스 안에 인스턴스 변수를 3개 이상 두지 말라는 조건이 있었고, 현재는 실무가 아니라 학습 중이기 때문에 최대한 조건에 맞추어 구현하는 것이 "수파리" 중 "수"에 해당한다고 느껴져서 이렇게 하였습니다.

csh0034 commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,39 @@
+    }
+
+    @Test
+    @DisplayName("답변 삭제 실패")
```
어떤 조건에서 실패하는지도 DisplayName 에 표현해주면 테스트코드를 이해할때 도움이 될 것 같습니다 🙂

csh0034 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,39 @@
+
+    @Test
+    @DisplayName("질문 삭제 실패")
+    void test2() {
```
bdd 스타일은 테스트를 이해하기 위함이므로 각 단계의 세팅이 없다면 꼭 주석을 추가해주시지 않아도 될 것 같습니다 😁

csh0034 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -1,15 +1,10 @@
-    @GeneratedValue
-    private Long id;
-
+public class DeleteHistory extends CreatedEntity {
```
하단 부분은 요구사항과 관련된 내용은 아니며 실제 구현을 한 작성자가
아니기 때문에 알수는 없지만 리팩토링을 할때 접근 방식에 대해서 고민해보면 좋을것 같아
코멘트 남깁니다.

createDate 에 대해서 Auditing 기능을 사용하도록 변경해주셨는데요
Auditing 기능과 동일한 기능인데 누락 한거라면 createAt 으로 추가했었을텐데
별도로 createDate 라는 다른 필드명으로 추가했다면 별도로 관리가 필요하기 때문이 아닐까라고
생각해볼수도 있을것 같습니다. 

실제 현업에서 레거시 코드를 리팩토링을 할때 해당 코드의 작성자가 없는 경우도 있기 때문에
리팩토링전에 기존 코드의 구현 의도를 파악해보려고 고민해보는것도 좋을것 같습니다!

csh0034 reviewed (changes requested):
안녕하세요.
볼링 미션을 함께 하게된 최승훈입니다. 

1단계 구현 관련하여 몇가지 코멘트 남겨두었는데 
확인해주시고 다시 리뷰 요청 부탁드립니다!

미션을 진행하면서 궁금하신 사항 있으시면
편하게 DM 주세요 😄

csh0034 commented:
질문에 대해서 답변드립니다.

Question 생성후 addAnswer 를 통해 Answer 를 추가해줄수 있습니다.

기존 코드에서 단위테스트가 어려운 이유는 트랜잭션 스크립트 패턴으로서
service 메서드에서 절차지향적으로 구현되어있기 때문에 answer 에 대해 
owner 를 체크하는 부분을 테스트 해야할 경우에 테스트하려는 부분 상단의 모든
조건을 만족해야만 테스트가 가능합니다.

따라서 실제 테스트 하기 위해 다른 코드들이 세팅 되어야만 하는 기존 구조를
각 객체(도메인)에게 검증과 행위를 부여함으로써 도메인 영역의 pojo 객체를
테스트 할수있는 구조로 리팩토링 하는것이 이번 단계의 목적이며 단위테스트가 어려운 
구조를 단위테스트가 가능한 구조로 바꾸는것입니다.

리뷰 및 답변에 대해 이해안가시는 부분이 있으시면 DM 주세요 😃

kangbada0728 committed:
refactor: Stream 적용 및 CannotDeleteException을 RuntimeException으로 변경

https://github.com/next-step/java-bowling/pull/922#discussion_r1014982533

kangbada0728 committed:
refactor: equals, hashcode, toString 재정의 및 공통 클래스를 추상클래스로 변경

https://github.com/next-step/java-bowling/pull/922#discussion_r1014984988
https://github.com/next-step/java-bowling/pull/922#discussion_r1014987566

kangbada0728 committed:
refactor: 실패 테스트의 조건을 DisplayName에 명시

https://github.com/next-step/java-bowling/pull/922#discussion_r1014991993

kangbada0728 committed:
refactor: 필요없는 주석 삭제

https://github.com/next-step/java-bowling/pull/922#discussion_r1014993525

kangbada0728 committed:
test: 질문과 답변 삭제 성공, 실패 테스트

https://github.com/next-step/java-bowling/pull/922#issuecomment-1305091795

csh0034 reviewed (approved):
코멘트 잘 반영해주셨네요 👍 
다음단계 진행해주세요!
